shoop/admin/modules/media/static_src/media/browser/FileUpload.js

```diff
@@ -6,69 +6,145 @@
+const _ = require("lodash");
+const remote = require("./util/remote");
+
+const queue = [];
+const queueCompleteCallbacks = [];
+var queueStatusDiv = null;
+
+function queueView() {
+    var className = "empty";
+    if(queue.length >= 0) {
+        className = (
+            _.all(queue, (file) => (file.status === "done" || file.status === "error")) ?
+                "done" : "busy"
+        );
+    }
+
+    return m("div.queue-view." + className, _.map(queue, (file) => {
+        return m("div.queue-file." + file.status, [
+            m("div.qf-name", file.name),
+            m("div.qf-progress", {style: "width: " + (file.progress || 0) + "%"})
+        ]);
+    }));
+}
+
+function updateQueueView() {
+    if(queueStatusDiv === null) {
+        if(queue.length === 0) {
+            return;  // Don't bother setting up the div if we're not doing anything actually
+        }
+        queueStatusDiv = document.createElement("div");
+        queueStatusDiv.id = "queue-status-ctr";
+        document.body.appendChild(queueStatusDiv);
+        // Yes, we're throwing away the ctrl instance; we don't need it
+        // and eslint would kvetch about it otherwise :)
+        m.mount(queueStatusDiv, {view: queueView, controller: _.noop});
+    }
+    m.redraw();  // XXX: It would be nice if we could redraw only one Mithril view...
 }
 
+const updateQueueViewSoon = _.debounce(updateQueueView, 50);
+
+function handleFileXhrComplete(xhr, file, error) {
+    if (xhr.status >= 400) {
+        error = true;
+    }
+    if (error) {
+        file.status = "error";
+
+    } else {
+        file.status = "done";
+        file.progress = 100;
+    }
+    setTimeout(processQueue, 50);  // Continue soon.
+    var messageText = null;
+    try {
+        const responseJson = JSON.parse(xhr.responseText);
+        if (responseJson && responseJson.message) {
+            messageText = responseJson.message;
+        }
+    } catch (e) {
+        console.log(e); // invalid JSON? pffff.
+    }
+    if (window.Messages) {
+        const response = {
+            error: (error ? "Error: " + file.name + ":" + messageText : null),
+            message: (!error ? messageText || "Uploaded: " + file.name : null)
+        };
+        remote.handleResponseMessages(response);
+    }
+}
 
-function createMessageShowingDoneFunction(done) {
-    return function(file, queueLength) {
-        if(window.Messages) Messages.enqueue({tags: "success", text: "Uploaded: " + file.name});
-        done();
+function beginUpload(file) {
+    if (file.status !== "new") {  // Already uploaded? Huh.
+        return false;
+    }
+    file.progress = 0;
+    file.status = "uploading";
+
+    const formData = new FormData();
+    formData.append("file", file.nativeFile);
+    const xhr = new XMLHttpRequest();
+    xhr.open("POST", file.url);
+    xhr.setRequestHeader("X-CSRFToken", window.ShoopAdminConfig.csrf);
+    xhr.onreadystatechange = function() {
+        if (xhr.readyState !== 4) {
```

Comment:
maybe add comment what this magical `4` is for future readers?
`// 4: request finished and response is ready`